### PR TITLE
HPCC-27438 Refactor lookup() functions to take an access mode parameter

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
NOTE: This should not change any behaviour.  It should only refactor the code and does not attempt to fix any problems that have been spotted.
Also requires https://github.com/hpcc-systems/LN/pull/1005

**Comment by ghalliday** [comment]
Fix released for the regression.  Cause was horrible.

    return write;

was returning true because it was casting the function to boolean and returning true.

**Review by richardkchapman — APPROVED** [review]

**Review by jakesmith — APPROVED** [review]
@ghalliday - looks good to me.

File: esp/clients/ws_dfsclient/ws_dfsclient.cpp [hunk 1]
```diff
@@ -861,6 +861,7 @@ class CLocalOrDistributedFile: implements ILocalOrDistributedFile, public CInter
             lfn.allowOsPath(true);
         if (!lfn.setValidate(fname))
             return false;
+        bool write = isWrite(accessMode);
```
Comment (richardkchapman) [inline esp/clients/ws_dfsclient/ws_dfsclient.cpp:864]: Any way to be sure that there are no other similar undetected issues introduced by this PR?

File: dali/daliadmin/daadmin.cpp [hunk 2]
```diff
@@ -1917,16 +1917,16 @@ void dfsreplication(const char *clusterMask, const char *lfnMask, unsigned redun
...
 void holdlock(const char *logicalFile, const char *mode, IUserDescriptor *userDesc)
 {
-    bool write;
+    AccessMode accessMode;
     if (strieq(mode, "read"))
-        write = false;
+        accessMode = AccessMode::tbdRead;
     else if (strieq(mode, "write"))
-        write = true;
+        accessMode = AccessMode::tbdWrite;
```
Comment (richardkchapman) [inline dali/daliadmin/daadmin.cpp:1924]: There's a reference to write below that may be incorrect

**Review by richardkchapman — CHANGES_REQUESTED** [review]
Maybe another write taht references the function?

File: dali/daliadmin/daadmin.cpp (hunk 2 above)
Comment (ghalliday) [reply dali/daliadmin/daadmin.cpp:1924]: well spotted.  I'll recheck all the examples/functions.

File: esp/clients/ws_dfsclient/ws_dfsclient.cpp (hunk 1 above)
Comment (ghalliday) [reply esp/clients/ws_dfsclient/ws_dfsclient.cpp:864]: I checked through again for every function that had a wait parameter.  Something to be wary of in the future.

**Review by richardkchapman — APPROVED** [review]

**Pull request merged**